src/live-programming.rst

```diff
@@ -20,9 +20,203 @@ Data Analysis Environments
+
+  This paper provided a theory of programmer navigation when debugging
+  based on information foraging theory.
+  (add sentence elaborating on connection to ift)
+  (add sentence about user study)
+  (add sentence about providing first principled design guidelines for making interactive debuggers)
+
+
+
 Interactive Debuggers
 ---------------------
 
+describe history of transition from physical computing
+systems to observing core dumps
+to more interactive stepping and navigation, graphical
+projections of execution behavior onto source code, etc
+
+Interactive Breakpoints
+^^^^^^^^^^^^^^^^^^^^^^^
+.. todo::
+  bibliography on breakpoints
+
+Time-Travel Debugging
+^^^^^^^^^^^^^^^^^^^^^
+.. todo::
+  bibliography on time-travel debugging,
+  distinguishing between record-replay and
+  reversible/omniscient debugging
+
+Program Slicing
+^^^^^^^^^^^^^^^
+
+(add description of program slicing as decomposition of programs)
+(provide some overview of different types of program slicing, e.g., static vs dynamic)
+
+.. container:: bib-item
+
+  .. bibliography:: debugging.bib
+    :filter: key == 'weiser1981slicing'
+
+  This paper introduced program slicing.
+  (add details about static slices)
+
+.. container:: bib-item
+
+  .. bibliography:: debugging.bib
+    :filter: key == 'ko2008whyline'
+
+  (describe connection to earlier work on debugging theory)
+  (describe the tool and how it uses dynamic program slicing)
+  (describe user study and result)
+
+.. container:: bib-item
+
+  .. bibliography:: debugging.bib
+    :filter: key == 'latoza2011reacher'
+
+  (describe connection to earlier work on reachability questions)
+  (describe the tool)
+  (describe user study and results)
+  (should probably move this out of program slicing)
+
+.. container:: bib-item
+
+  .. bibliography:: debugging.bib
+    :filter: key == 'perera2012explain'
+
+  (static and dynamic program slicing for functional programs)
+  (no implementation or evaluation)
+
+
+Algorithmic Debugging
+^^^^^^^^^^^^^^^^^^^^^
+
+*Algorithmic debugging* (also called *declarative debugging*) is a semi-automatic
+debugging technique in which the debugger automatically generates a series of
+questions to which the programmer's answers guide the search toward the bug.
+The debugger constructs an *execution tree* (ET), a data structure representing a
+program execution, and traverses it using some search strategy, asking the
+programmer at each ET node whether the represented subcomputation is correct
+in order to determine the next step.
+This technique guarantees that, if the programmer answers all the questions correctly,
+the bug will eventually be found.
+
+Although algorithmic debugging can be applied in any language paradigm, it is most
+suited for declarative languages, e.g., pure functional languages.
+to determine whether an ET node for a pure functional program is correct, the
+programmer need only check that the return value of the corresponding, referentially
```

Comment:
"referentially transparent" -- you should define this if you want to use it, but I don't think you need to say it at all